sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala

```diff
@@ -108,14 +109,38 @@ private[csv] object CSVInferSchema {
   }
 
   private def tryParseDouble(field: String): DataType = {
-    if ((allCatch opt field.toDouble).isDefined) {
+    val doubleTry = allCatch opt field.toDouble
```

Comment:
If we encounter a field that is just a long integer value, it is best represented as Decimal. If it has fractions then it must be a double. You are right all numeric types without fractions can be decimal.